eng/common/scripts/job-matrix/README.md

```diff
@@ -0,0 +1,453 @@
+# Azure Pipelines Matrix Generator
+
+* [Usage in a pipeline](#usage-in-a-pipeline)
+* [Matrix config file syntax](#matrix-config-file-syntax)
+ * [Fields](#fields)
+    * [matrix](#matrix)
+    * [include](#include)
+    * [exclude](#exclude)
+    * [displayNames](#displaynames)
+    * [$IMPORT](#import)
+* [Matrix Generation behavior](#matrix-generation-behavior)
+    * [all](#all)
+    * [sparse](#sparse)
+    * [include/exclude](#includeexclude)
+    * [displayNames](#displaynames-1)
+    * [Filters](#filters)
+    * [NonSparseParameters](#nonsparseparameters)
+    * [Under the hood](#under-the-hood)
+* [Testing](#testing)
+
+
+This directory contains scripts supporting dynamic, cross-product matrix generation for azure pipeline jobs.
+It aims to replicate the [cross-product matrix functionality in github actions](https://docs.github.com/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#example-running-with-more-than-one-version-of-nodejs),
+but also adds some additional features like sparse matrix generation, cross-product includes and excludes, and programmable matrix filters.
+
+This functionality is made possible by the ability for the azure pipelines yaml to take a [dynamic variable as an input
+for a job matrix definition](https://docs.microsoft.com/azure/devops/pipelines/process/phases?view=azure-devops&tabs=yaml#multi-job-configuration) (see the code sample at the bottom of the linked section).
+
+## Usage in a pipeline
+
+In order to use these scripts in a pipeline, you must provide a config file and call the matrix creation script within a powershell job.
+
+For a single matrix, you can include the `eng/pipelines/templates/jobs/job-matrix.yml` template in a pipeline:
+
+```
+jobs:
+- template: /eng/pipelines/templates/jobs/job-matrix.yml
```

Comment:
Why isn't the yml file part of this PR?

Reply:
This file doesn't actually exist (I think it used to). I guess this would now end up being the `archetype-sdk-tests-generate.yml` file instead. I can add that file to this repo, but I'm thinking of making the common pipeline yaml be a separate migration step than the scripts (since there are other inconsistencies still like the job filename archetype-sdk-tests/archetype-sdk-integration/archetype-sdk-tests-jobs).

Reply:
The more I think about this, the yaml samples are always going to be incomplete in some way, because I need to replicate the parameters passed in from stages and also replicate a mock archetype-sdk-tests-jobs.yml file. It may be better to just include a completely standalone sample here.

Reply:
Actually I think I have a decent working example that does part of the migration. Will push.